iree/compiler/Conversion/LinalgToLLVM/ConvertToLLVM2.cpp

```diff
@@ -0,0 +1,359 @@
+#include "llvm/Support/raw_ostream.h"
+#include "mlir/Conversion/AffineToStandard/AffineToStandard.h"
+#include "mlir/Conversion/LinalgToLLVM/LinalgToLLVM.h"
+#include "mlir/Conversion/SCFToStandard/SCFToStandard.h"
+#include "mlir/Conversion/StandardToLLVM/ConvertStandardToLLVM.h"
+#include "mlir/Conversion/StandardToLLVM/ConvertStandardToLLVMPass.h"
+#include "mlir/Conversion/VectorToLLVM/ConvertVectorToLLVM.h"
+#include "mlir/Conversion/VectorToSCF/VectorToSCF.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Dialect/LLVMIR/LLVMTypes.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/Dialect/StandardOps/Transforms/Passes.h"
+#include "mlir/Dialect/Vector/VectorOps.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+namespace iree_compiler {
+
+namespace {
+
+static constexpr int kIndexPackedBuffer = 0;
+static constexpr int kIndexPushConstant = 1;
+static constexpr int kIndexWorkGroupId = 2;
+static constexpr int kIndexWorkGroupCount = 3;
+static constexpr int kIndexWorkGroupSize = 4;
+
+// Convert to an LLVMFuncOp form with LLVM types. This implements an ABI that is
+// compatible with IREE and which cannot be represented in std atm.
+// Since it cannot be represented in std, this is an IREE-specific conversion.
+// The conversion rewrites the argument-less and return-less function:
+//
+// ```
+//    func @foo() { ... }
+// ```
+//
+// into:
+//
+// ```
+//    llvm.func foo(%packed_buffer_args: !llvm.ptr<!llvm.ptr<i8>>,
+//                  %push_constant: !llvm.ptr<i32>,
+//                  workgroup_id[3]: !llvm.ptr<i32>,
+//                  workgroup_count[3]: !llvm.ptr<i32>,
+//                  workgroup_size[3]: !llvm.ptr<i32>)
+// ```
+class ConvertFunc : public ConvertToLLVMPattern {
+ public:
+  explicit ConvertFunc(MLIRContext *context, LLVMTypeConverter &converter)
+      : ConvertToLLVMPattern(mlir::FuncOp::getOperationName(), context,
+                             converter) {}
+
+  LogicalResult matchAndRewrite(
+      Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const override {
+    auto funcOp = cast<FuncOp>(op);
+    FunctionType fnType = funcOp.getType();
+    (void)fnType;
+    if (!funcOp.isPublic()) return failure();
+
+    // illegal FuncOp must have 0 inputs.
+    assert(fnType.getNumInputs() == 0 && fnType.getNumResults() == 0);
+
+    // func foo(%packed_buffer_args: !llvm.ptr<!llvm.ptr<i8>>,
+    //          %push_constant: !llvm.ptr<i32>,
+    //          workgroup_id[3]: !llvm.ptr<i32>,
+    //          workgroup_count[3]: !llvm.ptr<i32>,
+    //          workgroup_size[3]: !llvm.ptr<i32>)
+    TypeConverter::SignatureConversion signatureConverter(/*numOrigInputs=*/0);
+    MLIRContext *context = rewriter.getContext();
+    auto packedBuffersArgsTy = LLVM::LLVMPointerType::get(
```

Comment:
Yes this was copied from the existing impl.
I am not 100% sure whether it would be transparent re the ABI or whether something weird would happen.
I'll investigate tomorrow.

Reply:
Updated to `!llvm.ptr<!llvm.array<3xi32>>`, I am unclear whether this will end up being better or worse but it passes the test.